### ISSUE TITLE
Environment Setup Plan

### ISSUE THREAD

**Issue by MusabAM** [MusabAM/mercata94-sub#1]
# NIDS Deep Learning & Quantum ML Environment Setup

## Overview

Setting up a comprehensive research environment for **Network Intrusion Detection System (NIDS)** using Deep Learning and Quantum Machine Learning methods.

### Hardware Profile
| Component | Specification |
|-----------|---------------|
| GPU | NVIDIA RTX 4050 (Notebook) |
| RAM | 32GB |
| CPU | Intel i7-13th Gen (H Processor) |
| OS | Windows |

---

## User Review Required

> [!WARNING]
> **Python Version Compatibility Issue**
> 
> Python 3.13 has limited support for some ML libraries as of December 2024:
> - **TensorFlow**: Only confirmed up to Python 3.12
> - **TensorFlow Quantum**: Only supports Python 3.10-3.12
> - **PyTorch**: Experimental support via nightly builds
> - **PennyLane**: ✅ Fully supports Python 3.13
>
> **Recommendation**: Use **Python 3.11** or **3.12** for maximum compatibility. Python 3.12 is recommended as the best balance between new features and library support.

> [!IMPORTANT]
> **CUDA Toolkit Required**
> 
> For GPU acceleration with RTX 4050, you'll need:
> - CUDA Toolkit 12.1+ (recommended: 12.4)
> - cuDNN 8.9+
> - Make sure NVIDIA drivers are up to date

---

### Project Structure

```
C:\Users\musab\.gemini\antigravity\scratch\nids-quantum-dl\
├── .vscode/                     # VS Code configuration
│   ├── settings.json
│   └── extensions.json
├── data/                        # Dataset storage
│   ├── raw/                     # Original datasets
│   │   ├── nsl-kdd/
│   │   ├── cicids2017/
│   │   ├── cicids2018/
│   │   └── unsw-nb15/
│   └── processed/               # Preprocessed data
├── notebooks/                   # Jupyter notebooks
│   ├── 01_data_exploration/
│   ├── 02_classical_dl/
│   ├── 03_quantum_dl/
│   └── 04_experiments/
├── src/                         # Source code
│   ├── __init__.py
│   ├── data/                    # Data loading & preprocessing
│   │   ├── __init__.py
│   │   ├── download.py
│   │   ├── preprocessing.py
│   │   └── datasets.py
│   ├── models/                  # Model architectures
│   │   ├── __init__.py
│   │   ├── classical/           # TensorFlow & PyTorch models
│   │   │   ├── cnn.py
│   │   │   ├── lstm.py
│   │   │   ├── transformer.py
│   │   │   └── autoencoder.py
│   │   └── quantum/             # Quantum ML models
│   │       ├── pennylane_models.py
│   │       └── tfq_models.py
│   ├── training/                # Training utilities
│   │   ├── __init__.py
│   │   ├── trainer.py
│   │   └── callbacks.py
│   ├── evaluation/              # Evaluation & metrics
│   │   ├── __init__.py
│   │   ├── metrics.py
│   │   └── visualization.py
│   └── utils/                   # Utility functions
│       ├── __init__.py
│       ├── config.py
│       └── logger.py
├── configs/                     # Configuration files
│   ├── datasets.yaml
│   ├── models.yaml
│   └── training.yaml
├── scripts/                     # Utility scripts
│   ├── download_datasets.py
│   ├── train.py
│   └── evaluate.py
├── tests/                       # Unit tests
├── results/                     # Experiment results
│   ├── models/                  # Saved models
│   ├── logs/                    # Training logs
│   └── figures/                 # Generated plots
├── requirements.txt             # Python dependencies
├── requirements-quantum.txt     # Quantum-specific dependencies
├── setup.py                     # Package setup
├── .env.example                 # Environment variables template
├── .gitignore
└── README.md
```

---

#### [NEW] [requirements.txt](file:///C:/Users/musab/.gemini/antigravity/scratch/nids-quantum-dl/requirements.txt)

Core ML/DL dependencies including:
- TensorFlow 2.15+ with CUDA support
- PyTorch 2.1+ with CUDA 12.1
- NumPy, Pandas, Scikit-learn
- Data visualization (Matplotlib, Seaborn, Plotly)
- Network traffic analysis tools
- Jupyter ecosystem

#### [NEW] [requirements-quantum.txt](file:///C:/Users/musab/.gemini/antigravity/scratch/nids-quantum-dl/requirements-quantum.txt)

Quantum ML dependencies:
- PennyLane with PyTorch & TensorFlow interfaces
- TensorFlow Quantum
- Cirq (Google's quantum computing framework)

---

#### [NEW] [.vscode/settings.json](file:///C:/Users/musab/.gemini/antigravity/scratch/nids-quantum-dl/.vscode/settings.json)

Configured for:
- Python interpreter pointing to virtual environment
- Auto-formatting with Black
- Linting with Pylint and Flake8
- Jupyter notebook settings

#### [NEW] [.vscode/extensions.json](file:///C:/Users/musab/.gemini/antigravity/scratch/nids-quantum-dl/.vscode/extensions.json)

Recommended extensions:
- Python, Pylance, Jupyter
- GitLens, GitHub Copilot
- YAML, Markdown support

---

#### [NEW] [src/data/download.py](file:///C:/Users/musab/.gemini/antigravity/scratch/nids-quantum-dl/src/data/download.py)

Dataset download scripts for:
- **NSL-KDD**: Classic benchmark dataset
- **CICIDS2017**: Modern labeled dataset with various attack types
- **CICIDS2018**: Extended version with additional attacks
- **UNSW-NB15**: Australian dataset with 9 attack families

#### [NEW] [src/data/preprocessing.py](file:///C:/Users/musab/.gemini/antigravity/scratch/nids-quantum-dl/src/data/preprocessing.py)

Preprocessing utilities:
- Feature engineering for network traffic
- Label encoding (binary & multiclass)
- Data normalization (MinMax, Standard, Robust)
- Train/validation/test splitting
- Handling class imbalance (SMOTE, undersampling)

---

#### Classical Models (TensorFlow & PyTorch)

| Model | File | Use Case |
|-------|------|----------|
| CNN | `src/models/classical/cnn.py` | Spatial feature extraction |
| LSTM | `src/models/classical/lstm.py` | Sequential pattern detection |
| Transformer | `src/models/classical/transformer.py` | Attention-based classification |
| Autoencoder | `src/models/classical/autoencoder.py` | Anomaly detection |

#### Quantum Models

| Model | File | Framework |
|-------|------|-----------|
| Variational Quantum Classifier | `src/models/quantum/pennylane_models.py` | PennyLane |
| Hybrid Quantum-Classical | `src/models/quantum/tfq_models.py` | TensorFlow Quantum |

---

### Notebooks

| Notebook | Purpose |
|----------|---------|
| `01_nsl_kdd_exploration.ipynb` | NSL-KDD data analysis |
| `02_cicids_exploration.ipynb` | CICIDS2017/2018 deep dive |
| `03_unsw_nb15_exploration.ipynb` | UNSW-NB15 analysis |
| `04_classical_baseline.ipynb` | Traditional ML baselines |
| `05_deep_learning_models.ipynb` | CNN, LSTM, Transformer training |
| `06_quantum_pennylane.ipynb` | PennyLane quantum experiments |
| `07_quantum_tfq.ipynb` | TensorFlow Quantum experiments |
| `08_model_comparison.ipynb` | Performance comparison across all models |

---

### Automated Tests

```bash
# 1. Verify Python environment
python --version  # Should show 3.11 or 3.12

# 2. Test CUDA/GPU availability
python -c "import torch; print(f'PyTorch CUDA: {torch.cuda.is_available()}')"
python -c "import tensorflow as tf; print(f'TensorFlow GPU: {tf.config.list_physical_devices(\"GPU\")}')"

# 3. Test quantum libraries
python -c "import pennylane as qml; print(f'PennyLane version: {qml.__version__}')"
python -c "import tensorflow_quantum as tfq; print('TensorFlow Quantum loaded successfully')"

# 4. Test data loading
python -c "from src.data import datasets; print('Data module loaded')"

# 5. Run unit tests
pytest tests/ -v
```

### Manual Verification

1. **Open project in VS Code** - Verify extensions load correctly
2. **Run sample notebook** - Confirm Jupyter integration works
3. **Download sample dataset** - Test download script functionality
4. **Train a simple model** - Verify GPU utilization with `nvidia-smi`

---

## Setup Instructions (for README)

```bash
# 1. Create virtual environment (use Python 3.11 or 3.12)
python -m venv venv
.\venv\Scripts\activate  # Windows

# 2. Upgrade pip
pip install --upgrade pip

# 3. Install core dependencies
pip install -r requirements.txt

# 4. Install quantum dependencies (optional, may have conflicts)
pip install -r requirements-quantum.txt

# 5. Verify installation
python scripts/verify_setup.py

# 6. Download datasets
python scripts/download_datasets.py --dataset cicids2017
```

---

## Next Steps After Approval

1. Create the project directory structure
2. Generate all configuration and source files
3. Create setup verification script
4. Generate starter notebooks
5. Create comprehensive README with usage examples
